File: alds1/ocaml/alds1_7_c.ml
wopen Printf
open Scanf

type node = { p : int option; l : int option; r : int option }
              
let id x = x

let find_root tr =
  let rec iter i = match tr.(i) with
      {p = Some j} -> iter j
    | _ -> i
  in iter 0

let preorder tr n =
  let rec iter ls = function
      None -> ls
    | Some i -> let ts1 = iter ls tr.(i).r in
                let ts2 = iter ts1 tr.(i).l in i :: ts2
  in iter [] (Some n)

let inorder tr n =
  let rec iter ls = function
      None -> ls
    | Some i -> let ts1 = iter ls tr.(i).r in
                let ts2 = iter (i :: ts1) tr.(i).l in ts2
  in iter [] (Some n)

let postorder tr n =
  let rec iter ls = function
      None -> ls
    | Some i -> let ts1 = iter (i::ls) tr.(i).r in
                let ts2 = iter ts1 tr.(i).l in ts2
  in iter [] (Some n)

let () =
  let n = scanf "%d\n" id in
  let btree = Array.make n {p = None; l = None; r = None} in
  let rec read n =
    if n = 0 then ()
    else begin
        let i, l, r = scanf "%d %d %d\n" (fun x y z -> (x, y, z)) in
        begin
          btree.(i) <- {btree.(i) with l = (if l <> (-1) then Some l else None);
                                       r = (if r <> (-1) then Some r else None)};
          if l <> (-1) then btree.(l) <- {btree.(l) with p = Some i};
          if r <> (-1) then btree.(r) <- {btree.(r) with p = Some i}
        end;
        read (n-1)
      end
  in
  read n;
  let r = find_root btree in
  printf "Preorder\n";
  preorder btree r |> List.iter (fun x -> printf " %d" x); printf "\n";
  printf "Inorder\n";
  inorder btree r |> List.iter (fun x -> printf " %d" x); printf "\n";
  printf "Postorder\n";
  postorder btree r |> List.iter (fun x -> printf " %d" x); printf "\n"

File: alds1/haskell/alds1_7_c.hs
import Control.Applicative ((<$>))
import Control.Monad (replicateM)
import Data.List (foldl')
import Data.Vector (Vector, (!), (//))
import qualified Data.Vector as V

main :: IO ()
main = do
  n <- readLn
  xs <- replicateM n (map read <$> words <$> getLine)
  solve n xs

solve :: Int -> [[Int]] -> IO ()
solve n xs = do putStrLn "Preorder"
                printList $ preOrder bt
                putStrLn "Inorder"
                printList $ inOrder bt
                putStrLn "Postorder"
                printList $ postOrder bt
  where ti = mkti n xs
        bt = mkBTree ti (findRoot ti)

mkti :: Int -> [[Int]] -> Vector (Maybe Int, Maybe Int, Maybe Int)
mkti n = foldl' f (V.replicate n (Nothing, Nothing, Nothing))
  where f v [i, l, r] = let x = let (bp, _, _) = v ! i in [(i, (bp, if l == (-1) then Nothing else Just l, if r == (-1) then Nothing else Just r))]
                            y = if l == (-1) then [] else let (_, ll, lr) = v ! l in [(l, (Just i, ll, lr))]
                            z = if r == (-1) then [] else let (_, rl, rr) = v ! r in [(r, (Just i, rl, rr))]
                        in v // (x ++ y ++ z)

findRoot :: Vector (Maybe Int, Maybe Int, Maybe Int) -> Maybe Int
findRoot = V.findIndex f
  where f (Nothing, _, _) = True
        f _ = False

data BTree a = Leaf | Node a (BTree a) (BTree a) deriving Show

mkBTree :: Vector (Maybe Int, Maybe Int, Maybe Int) -> Maybe Int -> BTree Int
mkBTree ti rt = f rt
  where f (Just i) = let (_, l, r) = ti ! i in Node i (f l) (f r)
        f Nothing = Leaf

foldPre :: (a -> b -> b) -> b -> BTree a -> b
foldPre f z Leaf = z
foldPre f z (Node x l r) = let y1 = foldPre f z r
                               y2 = foldPre f y1 l
                           in f x y2

foldIn :: (a -> b -> b) -> b -> BTree a -> b
foldIn f z Leaf = z
foldIn f z (Node x l r) = let y1 = foldIn f z r
                              y2 = f x y1
                          in foldIn f y2 l

foldPost :: (a -> b -> b) -> b -> BTree a -> b
foldPost f z Leaf = z
foldPost f z (Node x l r) = let y1 = f x z
                                y2 = foldPost f y1 r
                            in foldPost f y2 l

preOrder :: BTree a -> [a]
preOrder = foldPre (:) []

inOrder :: BTree a -> [a]
inOrder = foldIn (:) []

postOrder :: BTree a -> [a]
postOrder = foldPost (:) []

printList :: (Show a) => [a] -> IO ()
printList = putStrLn . (' ':) . unwords . map show
